examples/isomorphisms.dx

```diff
@@ -0,0 +1,258 @@
+'# Isomorphisms in Dex
+
+'`Iso a b` is the type of an isomorphism between `a` and `b`.
+
+:t MkIso
+> ((a:Type) ?-> (b:Type) ?-> {bwd: b -> a & fwd: a -> b} -> Iso a b)
+
+'This is a normal ADT, and you can construct your own isomorphisms.
```

Comment:
We should really make it a multi parameter type class... (Not to be fixed in this PR, just a general observation that I would like to have those implemented)

Reply:
As in `Iso` should be a typeclass? I'm not sure that would always do what we want, since there can be more than one isomorphism between two types, and I think it's not always obvious which isomorphism a user would want.

(Also, I thought Dex already had multi-parameter typeclasses? At least, I didn't see any logic that was specific to one parameter. But I don't think we've actually used that functionality)

Reply:
Type classes shouldn't have to be unique! This is a restriction of Haskell, but it's nothing fundamental. For example Idris relaxes it.

The general idea is that _sometimes_ those instances will be unique (e.g. there is only one isomorphism between `a & b` and `b & a`) and in those cases it seems silly to have to specify it explicitly. But there's no reason for letting people specify the exact instance in case it is actually ambiguous.

Reply:
🤯 

Interesting! I didn't know that about Idris. Sounds like something that could be cool to explore in the future.

I wonder, in what cases is it well defined to say there is only one isomorphism between `a & b` and `b & a`? It's certainly the case that there's only one polymorphic isomorphism (for all `a` and `b`) but for any particular `a` and `b` there might be more than one. Might be tricky to get dictionary synthesis to synthesize the right isomorphism when it's well-formed but not run into weird overlapping instance issues?

Reply:
Yeah, good point. There is one parametrically polymorphic isomoprhism and that's good enough for the type classes of Haskell, because they forbid ad-hoc specialization. I guess that in Dex you can do ad-hoc specialization, but if you attempt to define a refinement, the dictionary synthesis problem for that case will become ambiguous and inference should force you to pick one.